File: Java/StacksAndQueues/Problem3/TwoStacks.java
package StacksAndQueues.Problem3;

import java.util.ArrayList;
import java.util.List;

public class TwoStacks<T> {
    private List<T> stacks;
    private Integer i;
    private Integer j;

    public TwoStacks() {
        stacks = new ArrayList<>();
        stacks.add(null);
        stacks.add(null);
        i = 0;
        j = 1;
    }

    public void push1(T x) {
        stacks.set(i, x);
        if (i + 2 < stacks.size()) {
            i += 2;
        } else if (i + 1 < stacks.size()) {
            stacks.add(null);
            i += 2;
        } else {
            stacks.add(null);
            stacks.add(null);
            i += 2;
        }
    }

    public void push2(T x) {
        stacks.set(j, x);
        if (j + 2 < stacks.size()) {
            j += 2;
        } else if (j + 1 < stacks.size()) {
            stacks.add(null);
            j += 2;
        } else {
            stacks.add(null);
            stacks.add(null);
            j += 2;
        }
    }

    public T pop1() {
        if (i == 0) return null;
        T item = stacks.get(i - 2);
        i -= 2;
        stacks.set(i, null);
        return item;
    }

    public T pop2() {
        if (j == 1) return null;
        T item = stacks.get(j - 2);
        j -= 2;
        stacks.set(j, null);
        return item;
    }
}

File: Python/Practice Set 1/Stacks and Queues/two_stacks.py
# Problem link - https://www.geeksforgeeks.org/problems/implement-two-stacks-in-an-array/1

class TwoStacks:
    def __init__(self):
        self.stacks = [None, None]
        self.i, self.j = 0, 1

    def push1(self, x):
        self.stacks[self.i] = x
        if self.i + 2 not in range(len(self.stacks)):
            self.stacks.append(None)
            self.stacks.append(None)
        self.i += 2

    def push2(self, x):
        self.stacks[self.j] = x
        if self.j + 2 not in range(len(self.stacks)):
            self.stacks.append(None)
            self.stacks.append(None)
        self.j += 2

    def pop1(self):
        if self.i == 0:
            return
        self.i -= 2
        item = self.stacks[self.i]
        self.stacks[self.i] = None
        return item

    def pop2(self):
        if self.j == 0:
            return
        self.j -= 2
        item = self.stacks[self.j]
        self.stacks[self.j] = None
        return item


s = TwoStacks()
s.push1(2)
s.push1(3)
s.push2(4)
print(s.pop1())
print(s.pop2())
print(s.pop2())

s = TwoStacks()
s.push1(1)
s.push2(2)
print(s.pop1())
s.push1(3)
print(s.pop1())
print(s.pop1())

s = TwoStacks()
s.push1(2)
s.push1(3)
s.push1(4)
print(s.pop2())
print(s.pop2())
print(s.pop2())
